### ISSUE TITLE
Future plans for laz-perf

### ISSUE THREAD

**Issue by mccarthyryanc** [hobu/laz-perf#43]
I been watching some of the recent pull requests and merges from @tmontaigu (really awesome BTW!) and also noticed that you've had a lot of recent work to add support to read from buffers in [laspy](https://github.com/laspy/laspy/issues/78). Do you think you'll try add laz-perf to laspy to get native LAZ io?

@hobu , also noticed you had a pull request from a few years ago to do just this, but it doesn't look like the laspy devs went anywhere with it...

**Comment by hobu** [comment]
> @hobu , also noticed you had a pull request from a few years ago to do just this, but it doesn't look like the laspy devs went anywhere with it...

I am a lazpy developer 😄. I made a push to get laspy updated to support LAZ with lazperf, but I fizzled out of energy to do it, and I didn't want to integrate something that was incomplete. My branch at https://github.com/hobu/laspy/tree/lazperf-decomp was where I left off.

lazperf still needs to be caught up to LASzip 1.4, which it hasn't yet. It shouldn't be too difficult to do this, but it will take a chunk of time.

**Comment by tmontaigu** [comment]
> recent work to add support to read from buffers in laspy

Well I do have a branch that works, however as I replied in the laspy's issue it's not a clean solution, and 
I'm not a fan of the design of the code, so I'm not sur what to do

> Do you think you'll try add laz-perf to laspy to get native LAZ io?

I think I'll probably try something, I got lazperf working in my own las file reader so I know how to use it.

As for the laz-perf python bindings, I'll still need to make another PR, as there is bug at the moment
(the written chunk table at the end of the point data is not correct) and I want the api to feel less awkward  and more Pythonic

**Comment by hobu** [comment]
> I'm not a fan of the design of the code, so I'm not sur what to do

Feel free to rip things apart and put them back together again in both laspy and lazperf-python. laspy wasn't designed with the expectation that we would also be doing LAZ, and I find the design itself a bit overwrought (there's 2 too many levels of indirection that aren't really needed). As long as the tests pass, 👌.

**Comment by tmontaigu** [comment]
Well, I tried to refactor the code but I did not continue as it was getting hard, and I wasn't sure how the laspy maintainers would react to the massive changes that I would have done, because if I were to refactor laspy's code i would change ~65% of the code.

What I don't like in laspy's implementation is as you said the overthinking but also the fact that object-oriented techniques like inheritance are not used to handle cleanly the different file formats (mainly between [1.1, 1.2 ,1.3] files and 1.4 files).
 But to me the worst things are the circular dependencies between the classess in laspy which makes the code very hard to follow and even harder to modify.

thats why I ended up starting something from scratch in the mean time

**Comment by hobu** [comment]
> Well, I tried to refactor the code but I did not continue as it was getting hard, and I wasn't sure how the laspy maintainers would react to the massive changes that I would have done, because if I were to refactor laspy's code i would change ~65% of the code.

Speaking as a laspy maintainer, I would tell you not to feel bashful about doing this. I've felt that the laspy code is a bit too complex, and things like LAZ support are an opportunity to flush out things that aren't needed. My attempts to plumb things in were met with the same challenges you describe, but I didn't have three days to spend to go and fix it right while modernizing the code to today's Python idioms at the same time. If you have that time, go for it. We'll merge it.

**Comment by grantbrown** [comment]
@hobu I totally agree.

**Comment by kbevers** [comment]
I also agree. Go ahead and improve laspy! None of us really spend much time on it so any contribution (that doesn't completely ruing things) is valuable. This sounds like an awesome contribution.

**Comment by tmontaigu** [comment]
Then I will give the refactoring another shot.

I want laspy to work on streams/buffer and not just files present on the disk as I was able to quickly make it work in my [pylas](https://github.com/tmontaigu/pylas)
As consequence, I will have to remove the mmap. Using a mmap by default does not make sense if we want a proper LAZ support. However I think i'll reintroduce the mmap as an optionnal way to read LAS data because it's still a cool way to handle large uncompressed files.

There is also  all the different file modes that I will remove because it will make less sense to still have them and the current handling of it in the code just make things more complex.
And I want to follow the python principle "We are all consenting adults", so If you don't want a file to be modified don't modify it.

In the end there are things that will inevitably have to change in the end user interface, and I don't know if I will achieve this refactoring

**Comment by kbevers** [comment]
> In the end there are things that will inevitably have to change in the end user interface, and I don't know if I will achieve this refactoring

If you find it necessary to change the API I'd say that is worth it to have proper support for laz and streams. No objections from me on that part. We can just call the next release Laspy 2.0 and be done with it.

**Comment by grantbrown** [comment]
I wholeheartedly agree with @hobu and @kbevers. Laspy started as a learning
exercise for both python and the LAS standard, so I have no objections to a
complete overhaul.

**Comment by tmontaigu** [comment]
@brycefrank From what I've tested a few months ago, stream access (as in read the points one by one)
was extremely slow in python, depending on how LAX files allows to filter points (do we just need the index, or the x,y,z), it may be done in a faster way

**Comment by brycefrank** [comment]
@tmontaigu You are correct. I was able to check this on Friday with the same result. I don't think there is a straightforward way around this problem. My solution for now is to just call `lasindex` from the operating system and parse the `.lax` file after it is produced (similar to how laspy calls `laszip`).

Here is my progress so far:
https://github.com/brycefrank/laxpy

**Comment by tmontaigu** [comment]
@brycefrank I don't know much about LAX files and what they allow us to do in terms of queries,
but if what we can get from the LAX file queries is a bunch of point indexes then we might be able to get something with acceptable performance in pure Python for LAS files.

**Comment by brycefrank** [comment]
@tmontaigu 

Let's continue this over [here](https://github.com/brycefrank/laxpy/issues/1) if you don't mind!

**Comment by Sajjadmanal** [comment]
@hobu @kbevers Could you please help me to read a .laz file in python? I am only able to read .las file using laspy.
